### PR TITLE
EQL: Introduce repeatable queries

### PR DESCRIPTION
**Pull request by costin**
Allow individual queries within a sequence to be repeated through a
dedicated keyword without having physical duplication.

sequence
  queryA [runs=2]
  queryB
  queryC [runs=3]
  queryD

is the same as:

sequence
  queryA
  queryA
  queryB
  queryC
  queryC
  queryC
  queryD

but more concise.

### CODE REVIEW

File: x-pack/plugin/eql/qa/common/src/main/resources/test_queries.toml [hunk 1]
```diff
@@ -733,7 +849,19 @@ sequence
...
 '''
-expected_event_ids  = [54, 55, 61, 67]
+expected_event_ids = [54, 55, 61, 67]
+
+[[queries]]
+name = "fourSequencesByPidAndProcessPathWithUntil-Repeat"
+query = '''
+sequence
+  [process where opcode == 1] by unique_pid,  process_path
+  [file where opcode == 0]    by unique_pid,  process_path  repeat=3
```
Comment (astefan) [inline x-pack/plugin/eql/qa/common/src/main/resources/test_queries.toml:859]: How about the following scenario?
```
[file where opcode == 0] by unique_pid, process_path repeat=2
[file where opcode == 0] by unique_pid, process_path repeat=1
```

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java [hunk 2]
```diff
@@ -251,7 +251,37 @@ public Sequence visitSequence(SequenceContext ctx) {
...
+                    throw new ParsingException(source(numberCtx), "A query cannot be repeated more than 100 times; found [{}]", value);
+                }
+                repeat = (int) value;
+            }
+
+            int numberOfQueries = queries.size() + repeat;
+            if (numberOfQueries > 256) {
+                throw new ParsingException(
+                    source(sequenceTermCtx),
+                    "Sequence cannot contains more than 256 queries; found [{}]",
```
Comment (astefan) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java:273]:
```suggestion
                    "Sequence cannot contain more than 256 queries; found [{}]",
```

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java [hunk 3]
```diff
@@ -251,7 +251,37 @@ public Sequence visitSequence(SequenceContext ctx) {
...
+                );
+            }
+
+            for (int i = 0; i < repeat; i++) {
+                queries.add(sequenceTerm);
+            }
+        }
+
+        if (queries.size() < 2) {
+            throw new ParsingException(source, "A sequence requires a minimum of 2  queries, found [{}]", queries.size());
```
Comment (astefan) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java:284]:
```suggestion
            throw new ParsingException(source, "A sequence requires a minimum of 2 queries, found [{}]", queries.size());
```

Comment (astefan) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java:284]: Can you also test this new exception?

**Review by astefan — COMMENTED** [review]
Looks great.
Some unit tests need to be added for the new exceptions. And the scenarios where multiple `repeat`s are used in the same sequence need clarification.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java [hunk 4]
```diff
@@ -253,7 +254,45 @@ public Sequence visitSequence(SequenceContext ctx) {
...
+            // check runs
+            Token key = sequenceTermCtx.key;
+            if (key != null) {
+                String k = key.getText();
+                if (RUNS.equals(k) == false) {
+                    throw new ParsingException(source(key), "Unrecognized option [{}], expecting [{}]", k, RUNS);
+                }
+            }
+
+            int runs = 1;
```
Comment (astefan) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java:266]: If this is limited to 100, why not using `byte`?

**Review by astefan — APPROVED** [review]
LGTM. Left one minor comment.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java (hunk 4 above)
Comment (costin) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java:266]: Habit - since it's a local variable it doesn't make a difference.

**Comment by rw-access** [comment]
what's up with the brackets? the `[runs=X]` syntax doesn't make sense (choice of word aside) and now the meaning of brackets is very overloaded and confusing.

before, it just had exactly one meaning that was used in a _lot_ of contexts. it wasn't on accident that it always contained an event filter: `[event category where condition]`?

some of those usages: terms in a `sequence` or `join`, `descendant of`, `until`, etc. could also be used to index into arrays but that's still distinct `field[1]`

also the placement (after `by`) does seem a little odd in choice.
i would expect this to be more intuitive (again, choice of word aside)
```
sequence
  [process where opcode == 1]          by unique_pid
  [file where    opcode == 0] runs=3   by unique_pid
```

**Comment by costin** [comment]
Hi Ross. Thanks for the feedback.

### tl;dr

To avoid stretching `[]`, it is being replaced with `with` a construct that already has the semantics we want and sounds good enough.

```
sequence
  [process where opcode == 1] by unique_pid
  [file where    opcode == 0] by unique_pid with runs=2
  [file where    opcode == 0] by unique_pid
```

### Long story

There were several discussions going back and forth on how to declare `runs` which is a meta property of the query. 
The overall goal in the syntax is to preserve the duplication without affecting the core query so that moving from one declaration to multiple and back is as easy as possible:

```
sequence
  [queryA] by X
  [queryA] by X
  [queryA] by X

// somehow becomes
sequence
  [queryA] by X // make it 3 times
```

This favors declaring the property as suffix, just like `by` and `with`. 
Per this PR, another goal was to minimize the amount of symbols so to avoid adding a new one (such as `{}`), `[]` was used which at the time seemed like a good idea (famous last words).

Thanks to your feedback we reconsidered the reuse of `[]` and opted for `with` since it fits a lot of the properties wanted and continues the style of using prepositions/_glue_ words as markers and block starters.
It keeps the same rules as within sequence declaration (needs to be as a suffix), allows enumeration for possible extensions in the future and allows other potential prepositions to be added further down the road:

```
// variants

[queryA]
[queryB] by X
[queryC] with runs=3
[queryD] by Z with runs=2
```

**Comment by rw-access** [comment]
I like it!
